kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/VectorUtils.java

```diff
@@ -117,27 +94,46 @@ public int getSize() {
 
       @Override
       public ColumnVector getKeys() {
-        return stringVector(keys);
+        return buildColumnVector(keys, StringType.STRING);
       }
 
       @Override
       public ColumnVector getValues() {
-        return stringVector(values);
+        return buildColumnVector(values, StringType.STRING);
+      }
+    };
+  }
+
+  /** Creates an {@link ArrayValue} from list of objects. */
+  public static ArrayValue buildArrayValue(List<?> values, DataType dataType) {
+    if (values == null) {
+      return null;
+    }
+    return new ArrayValue() {
+      @Override
+      public int getSize() {
+        return values.size();
+      }
+
+      @Override
+      public ColumnVector getElements() {
+        return buildColumnVector(values, dataType);
       }
     };
   }
 
   /**
-   * Utility method to create a {@link ColumnVector} for given list of strings.
+   * Utility method to create a {@link ColumnVector} for given list of object, the object should be
+   * primitive type or an Row instance.
    *
    * @param values list of strings
-   * @return a {@link ColumnVector} with the given values of type {@link StringType}
+   * @return a {@link ColumnVector} with the given values type.
    */
-  public static ColumnVector stringVector(List<String> values) {
+  public static ColumnVector buildColumnVector(List<?> values, DataType dataType) {
```

Comment:
Move the ColumnVector impl out of here into `io.delta.kernel.internal.data`. This is similar to the GenericRow we have, may be name this GenericColumnVector (which takes an array of value objects and exposes as ColumnVector)

Reply:
+1